core/src/main/scala/kafka/cluster/Partition.scala

```
@@ -817,6 +817,18 @@ class Partition(val topicPartition: TopicPartition,
       case None => localReplica.logEndOffset.messageOffset
     }
 
+    // Only actually check the HW if this is a "latest" offset client request
+    if (isolationLevel.isDefined && timestamp == ListOffsetRequest.LATEST_TIMESTAMP) {
+      // Throw if the HW has not caught up with the start offset from this epoch
+      leaderEpochStartOffsetOpt
+        .filter(leo => leo > localReplica.highWatermark.messageOffset)
+        .map(leo => Errors.OFFSET_NOT_AVAILABLE.exception(s"Failed to fetch offsets for " +
```

Could we change this `map` to `foreach` and throw from here?

If we are only checking this for the latest case, I guess we could move this into the `if` block below. However, I am considering whether we should check this for general timestamp queries. The problem is that we may be excluding some committed data from a query if we filter only based on the local high watermark. This might cause a timestamp which was visible to disappear temporarily after a leader change. 

So we could change the code slightly below. If `fetchOffsetByTimestamp` returns an offset that is larger than `lastFetchableOffset` and `leaderEpochStartOffsetOpt` is defined, then instead of returning `None`, we could raise `OFFSET_NOT_AVAILABLE`. Would that make sense?

---

I think that makes sense, +1

---

Ok, I incorporated this logic into the latest commit. I also reorganized the code to make it easier to understand (hopefully)